### PR TITLE
Support extracting functions from a parsed query

### PR DESCRIPTION
**Pull request by Tassosb**
Exposes 3 new public methods on `PgQuery` that extract functions from the parsed query following a similar pattern as the `#tables` methods.
- `#functions`
- `#ddl_functions`
- `#call_functions`

**Limitations:**
In Postgres multiple functions can have the same name as long as they have different argument types. This PR ignores that nuance, so functions with the same name are basically treated as the same function.

For example, if a statement creates  a function `foo(integer)` and calls a function `foo(varchar)`, we'd get `#ddl_functions == ['foo']` and `#call_functions == ['foo']`. Just looking at the return values, there is no way of knowing whether or not these are in fact the same functions.

When a function gets called, it seems hard to infer the argument types of the function and thereby uniquely identify the called function. I'm guessing there is a way but I'm not sure if it is supported by the parser.

Even with this limitation, this feature is still useful for me.

### CODE REVIEW

File: lib/pg_query/parse.rb [hunk 1]
```diff
@@ -48,27 +49,48 @@ def ddl_tables
...
+  # types.
+  def functions
+    functions_with_types.map { |f| f[:function] }
+  end
+
+  def ddl_functions
+    functions_with_types.select { |f| f[:type] == :ddl }.map { |f| f[:function] }
+  end
+
+  def call_functions
```
Comment (Tassosb) [inline lib/pg_query/parse.rb:63]: I tried to name these methods following the pattern of `select_tables`, `ddl_tables`...

**Comment by lfittl** [comment]
@Tassosb Nice!

As a quick FYI, I'm currently focused on rebasing pg_query to Postgres 13, which will involve a few internal changes as well.

Would you prefer if we merge this into the current branch, and I rebase the work to the internal changes, or would you like to do so after the Postgres 13 release is out? (depends a bit how heavily you are still modifying this code)

**Comment by Tassosb** [comment]
Oh nice! I'll wait to merge this until after the Postgres 13 release then. It's possible I may need to make more modifications. Should I close this PR for now?

**Comment by lfittl** [comment]
> Oh nice! I'll wait to merge this until after the Postgres 13 release then. It's possible I may need to make more modifications. Should I close this PR for now?

Sounds good! Feel free to leave it open, I don't have a preference either way :)

**Comment by lfittl** [comment]
@Tassosb The Postgres 13-based release is out now: https://pganalyze.com/blog/pg-query-2-0-postgres-query-parser (2.0.1 release)

Could you update your PR to the new parse tree format?

**Comment by Tassosb** [comment]
Nice release! I updated the PR to the new parse tree format.

**Review by lfittl — APPROVED** [review]
Looks good - thanks for the contribution!

**Pull request merged**